File: src/test/java/com/aether/interpreter/StringInterpTest.java
package com.aether.interpreter;

import static org.junit.jupiter.api.Assertions.assertEquals;

import com.aether.lexer.Scanner;
import com.aether.parser.Parser;
import com.aether.parser.ast.Stmt;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

/** Tests for string interpolation: "text ${expr} text". */
class StringInterpTest {

  private Evaluator evaluator;

  @BeforeEach
  void setUp() {
    evaluator = Evaluator.withoutStdlib();
  }

  private String eval(String source) {
    List<Stmt> stmts = new Parser(new Scanner(source).scanTokens()).parse();
    for (int i = 0; i < stmts.size() - 1; i++) {
      evaluator.execStmt(stmts.get(i));
    }
    if (!stmts.isEmpty() && stmts.getLast() instanceof Stmt.ExprStmt es) {
      return Builtins.display(evaluator.evalExpr(es.expr()));
    }
    if (!stmts.isEmpty()) {
      evaluator.execStmt(stmts.getLast());
    }
    return "null";
  }

  @Test
  void variable() {
    assertEquals("Hello Aether", eval("let name = \"Aether\"\n\"Hello ${name}\""));
  }

  @Test
  void number() {
    assertEquals("Value: 42", eval("let n = 42\n\"Value: ${n}\""));
  }

  @Test
  void arithmeticExpression() {
    assertEquals("Result: 6", eval("\"Result: ${1 + 2 + 3}\""));
  }

  @Test
  void multiple() {
    assertEquals("3 + 4 = 7", eval("let a = 3\nlet b = 4\n\"${a} + ${b} = ${a + b}\""));
  }

  @Test
  void noPlaceholder() {
    assertEquals("plain string", eval("\"plain string\""));
  }

  @Test
  void boolValue() {
    assertEquals("flag is true", eval("let flag = true\n\"flag is ${flag}\""));
  }

  @Test
  void nestedCall() {
    assertEquals("double(5) = 10",
        eval("fn double(x) { return x * 2 }\n\"double(5) = ${double(5)}\""));
  }

  @Test
  void atStart() {
    assertEquals("99 bottles", eval("let x = 99\n\"${x} bottles\""));
  }

  @Test
  void atEnd() {
    assertEquals("count: 3", eval("let x = 3\n\"count: ${x}\""));
  }

  @Test
  void onlyInterpolation() {
    assertEquals("42", eval("let x = 42\n\"${x}\""));
  }

  @Test
  void nullValue() {
    assertEquals("got null", eval("let x = null\n\"got ${x}\""));
  }

  @Test
  void arrayValue() {
    assertEquals("arr: [1, 2, 3]", eval("let a = [1, 2, 3]\n\"arr: ${a}\""));
  }

  @Test
  void stringValue() {
    assertEquals("say hi", eval("let s = \"hi\"\n\"say ${s}\""));
  }

  @Test
  void interpolationInLoop() {
    assertEquals("x=0 x=1 x=2",
        eval("let parts = []\n"
            + "let i = 0\n"
            + "while (i < 3) { parts.push(\"x=${i}\")\ni += 1 }\n"
            + "parts[0] + \" \" + parts[1] + \" \" + parts[2]"));
  }
}

File: tests/string_interp_test.rs
//! Tests for string interpolation

use aether_lang::interpreter::Evaluator;
use aether_lang::lexer::Scanner;
use aether_lang::parser::Parser;

fn run(source: &str) -> Result<String, String> {
    let mut scanner = Scanner::new(source);
    let tokens = scanner.scan_tokens().map_err(|e| e.to_string())?;
    let mut parser = Parser::new(tokens);
    let program = parser.parse().map_err(|e| e.to_string())?;
    let mut evaluator = Evaluator::new_without_stdlib();

    for stmt in &program.statements[..program.statements.len().saturating_sub(1)] {
        evaluator.exec_stmt(stmt).map_err(|e| e.to_string())?;
    }

    if let Some(last) = program.statements.last() {
        if let aether_lang::parser::ast::Stmt::Expr(expr) = last {
            let value = evaluator.eval_expr(expr).map_err(|e| e.to_string())?;
            return Ok(format!("{}", value));
        }
        evaluator.exec_stmt(last).map_err(|e| e.to_string())?;
    }

    Ok("null".to_string())
}

#[test]
fn test_interp_variable() {
    let result = run(r#"let name = "Aether"  "Hello ${name}""#);
    assert_eq!(result.unwrap(), "Hello Aether");
}

#[test]
fn test_interp_number() {
    let result = run(r#"let n = 42  "Value: ${n}""#);
    assert_eq!(result.unwrap(), "Value: 42");
}

#[test]
fn test_interp_expression() {
    let result = run(r#""Result: ${1 + 2 + 3}""#);
    assert_eq!(result.unwrap(), "Result: 6");
}

#[test]
fn test_interp_multiple() {
    let result = run(r#"let a = 3  let b = 4  "${a} + ${b} = ${a + b}""#);
    assert_eq!(result.unwrap(), "3 + 4 = 7");
}

#[test]
fn test_interp_no_placeholder() {
    let result = run(r#""plain string""#);
    assert_eq!(result.unwrap(), "plain string");
}

#[test]
fn test_interp_bool() {
    let result = run(r#"let flag = true  "flag is ${flag}""#);
    assert_eq!(result.unwrap(), "flag is true");
}

#[test]
fn test_interp_nested_call() {
    let source = r#"
fn double(x) { return x * 2 }
"double(5) = ${double(5)}"
"#;
    let result = run(source);
    assert_eq!(result.unwrap(), "double(5) = 10");
}

#[test]
fn test_interp_at_start() {
    let result = run(r#"let x = 99  "${x} bottles""#);
    assert_eq!(result.unwrap(), "99 bottles");
}

#[test]
fn test_interp_at_end() {
    let result = run(r#"let x = 7  "number ${x}""#);
    assert_eq!(result.unwrap(), "number 7");
}
